internal/wal_verify_handler.go

```diff
@@ -0,0 +1,219 @@
+		EndSegment:    sequence.maxSegmentNo.getFilename(sequence.timelineId),
+		Status:        status,
+		SegmentsCount: len(sequence.walSegmentNumbers),
+	}
+}
+
+// QueryCurrentWalSegment() gets start WAL segment from Postgres cluster
+func QueryCurrentWalSegment() WalSegmentDescription {
+	conn, err := Connect()
+	tracelog.ErrorLogger.FatalfOnError("Failed to establish a connection to Postgres cluster %v", err)
+
+	queryRunner, err := newPgQueryRunner(conn)
+	tracelog.ErrorLogger.FatalfOnError("Failed to initialize PgQueryRunner %v", err)
+
+	currentSegmentNo, err := getCurrentWalSegmentNo(queryRunner)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get current WAL segment number %v", err)
+
+	currentTimeline, err := getCurrentTimeline(conn)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get current timeline %v", err)
+
+	err = conn.Close()
+	tracelog.WarningLogger.PrintOnError(err)
+
+	// currentSegment is the current WAL segment of the cluster
+	return WalSegmentDescription{Timeline: currentTimeline, Number: currentSegmentNo}
+}
+
+// HandleWalVerify queries the current cluster WAL segment and timeline
+// and travels through WAL segments in storage in reversed chronological order (starting from that segment)
+// to find any missing WAL segments that could potentially fail the PITR procedure
+func HandleWalVerify(rootFolder storage.Folder, startWalSegment WalSegmentDescription, outputWriter WalVerifyOutputWriter) {
+	walFolder := rootFolder.GetSubFolder(utility.WalPath)
+	storageFileNames, err := getFolderFilenames(walFolder)
+	tracelog.ErrorLogger.FatalfOnError("Failed to get WAL folder filenames %v", err)
+
+	storageSegments := getSegmentsFromFiles(storageFileNames)
+	timelineSwitchMap, err := createTimelineSwitchMap(startWalSegment.Timeline, walFolder)
+	tracelog.ErrorLogger.FatalfOnError("Failed to initialize timeline history map %v", err)
+	walSegmentRunner := NewWalSegmentRunner(startWalSegment, storageSegments, 1, timelineSwitchMap)
+
+	// maxConcurrency is needed to determine max amount of missing WAL segments
+	// after the last found WAL segment which can be skipped ("uploading" segment sequence size)
+	maxConcurrency, err := getMaxUploadConcurrency()
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	segmentScanner := NewWalSegmentScanner(walSegmentRunner)
+	err = runWalIntegrityScan(segmentScanner, maxConcurrency)
+	tracelog.ErrorLogger.FatalfOnError("Failed to perform WAL segments scan %v", err)
+
+	integrityScanSegmentSequences, err := collapseSegmentsByStatusAndTimeline(segmentScanner.ScannedSegments)
+	tracelog.ErrorLogger.FatalOnError(err)
+
+	err = outputWriter.Write(newWalVerifyResult(integrityScanSegmentSequences))
+	tracelog.ErrorLogger.FatalOnError(err)
+}
+
+// runWalIntegrityScan invokes the following storage scan series:
+// 1. At first, it runs scan until it finds some segment in WAL storage
+// and marks all encountered missing segments as "missing, probably delayed"
```

Comment:
For example, we have 8 WAL segments locally and latest WAL segment in storage is 4'th.
So we mark 5,6,7,8 as Delayed